### PR TITLE
refactor(router): Add warning for `relativeLinkResolution: 'legacy'`

### PR DESCRIPTION
**Pull request by atscott**
This change adds code to compute the corrected value for a link,
regardless of the `relativeLinkResolution` value. Then, if the
`relativeLinkResolution` is set to `legacy` and differs from the correct
value, a warning is printed to the console in dev mode.

This change is meant to assist in notifying developers that they have
code which relies on the deprecated, broken behavior so they can fix and
update the code before the `relativeLinkResolution` option is fully
removed.

### CODE REVIEW

File: packages/router/src/create_url_tree.ts [hunk 1]
```diff
@@ -24,13 +24,30 @@ export function createUrlTree(
...
+  const result = createTreeUsingPathIndex(route.snapshot._lastPathIndex);
+
+  // Check if application is relying on `relativeLinkResolution: 'legacy'`
+  if (typeof ngDevMode === 'undefined' || !!ngDevMode) {
+    const correctedResult = createTreeUsingPathIndex(route.snapshot._correctedLastPathIndex);
+    if (correctedResult.toString() !== result.toString()) {
+      console.warn(
+          `relativeLinkResolution: 'legacy' is deprecated and will be removed in a future version of Angular. The link to ${
+              result.toString()} will change to ${
+              correctedResult.toString()} if the code is not updated before then.`);
```
Comment (AndrewKushnir) [inline packages/router/src/create_url_tree.ts:46]: The error looks great (it includes both old and new URLs), but I'm wondering if we can improve it even more. Currently, there is no indication for users where to look at (i.e. which component/template/directive). Can we supply a bit more info to the `createUrlTree` call from the RouterLink and may be refer to the original value used in the `[routerLink]` as well as mention that it's coming from the `[routerLink]` binding?

**Comment by atscott** [comment]
[presubmit](https://fusion2.corp.google.com/presubmit/tap/439357021/OCL:439357021:BASE:439313712:1649094550809:baa98911/targets)

File: packages/router/src/recognize.ts [hunk 2]
```diff
@@ -163,7 +165,9 @@ export class Recognizer {
           segments, params, Object.freeze({...this.urlTree.queryParams}), this.urlTree.fragment,
           getData(route), getOutlet(route), route.component!, route,
           getSourceSegmentGroup(rawSegment), getPathIndexShift(rawSegment) + segments.length,
-          getResolve(route));
+          getResolve(route),
+          (NG_DEV_MODE ? getCorrectedPathIndexShift(rawSegment) : getPathIndexShift(rawSegment)) +
```
Comment (AndrewKushnir) [inline packages/router/src/recognize.ts:169]: I'd propose adding a quick comment here (and in the code below where we use `NG_DEV_MODE`) to explain why the value is different depending on the `ngDevMode` flag.

Comment (AndrewKushnir) [reply packages/router/src/recognize.ts:169]: It looks like we already perform the `getPathIndexShift(rawSegment)` calculation above, we can probably save the result in a const and reuse here.

**Review by AndrewKushnir — COMMENTED** [review]
@atscott looks great 👍

File: packages/router/src/create_url_tree.ts (hunk 1 above)
Comment (atscott) [reply packages/router/src/create_url_tree.ts:46]: We can't do this directly without affecting the public API, which I think we should avoid. One option might be to throw and catch an error inside this function to get a stack of where the `createUrlTree` call is coming from. Not sure if that's a good idea... WDYT?

Comment (AndrewKushnir) [reply packages/router/src/create_url_tree.ts:46]: If there is no easy way to propagate this info - no worries, we can improve in the future if there is an additional feedback.

**Review by AndrewKushnir — APPROVED** [review]

**Comment by atscott** [comment]
merge assistance: Alex's review is not needed

**Comment by jessicajaniuk** [comment]
This PR was merged into the repository by commit d180db15bf7aef83335ec152002e58a1d9e3031e.

**Comment by jessicajaniuk** [comment]
This caused breakages in google3 and had to be rolled back.

**Comment by atscott** [comment]
[TGP](http://tap/439357021)

**Comment by atscott** [comment]
[TGP](https://fusion2.corp.google.com/presubmit/tap/439357021/OCL:439357021:BASE:441495601:1649865769795:100da9b8/targets) is green after rerun

merge assistance: this is a resubmit

**Comment by jessicajaniuk** [comment]
This PR was merged into the repository by commit 47e7a2850b6e51eeb455ebb6f54339fd2c71b6cd.

**Pull request closed without merging**